R/runner.R

```diff
@@ -226,11 +242,72 @@ orderly_runner_ <- R6::R6Class(
     #' Submit an arbitrary job on the queue
     #'
     #' @param job A quoted R expression.
-    #' @param environment Environment to run the expression in.
+    #' @param depends_on Task ids for any dependencies of this job.
     #'
     #' @return Task id
-    submit = function(job, environment = parent.frame()) {
-       self$queue$enqueue_(job, environment)
+    submit = function(expr, depends_on = NULL) {
+      self$queue$enqueue_(expr, depends_on = depends_on)
+    },
+
+    #' @description
+    #' Queue a workflow.
+    #'
+    #' @param reports Details of reports to be run.
+    #' @param ref The git sha to run the workflow.
+    #' @param changelog Description of changes to the reports - applied to
+    #' all reports.
+    #' @param poll How frequently to poll for the report ID being available.
+    #' Applied to each of the reports run in the workflow
+    #' @param timeout Timeout for each of the reports run as part
+    #' of the workflow default 3 hours.
+    #'
+    #' @return The key for the workflow and each individual report
+    submit_workflow = function(reports, ref = NULL, changelog = NULL,
+                               poll = 0.1, timeout = 60 * 60 * 3) {
+      private$assert_ref_switching_allowed(ref)
+      git_pull(self$root)
+      if (!is.null(ref)) {
+        git_fetch(self$alternative_root)
+        prev <- git_checkout_branch(ref, root = self$alternative_root)
+        on.exit(git_checkout_branch(prev, root = self$alternative_root))
+      } else {
+        git_pull(self$alternative_root)
```

Comment:
I am a bit concerned about this pull here. We can leave it in for now but it would be good to think about the points at which we actually do this refresh typically. By the time we've hit this point we've already discussed with the front end the potentially missing reports I think, so there should not be need to update it again?

Reply:
Yes thanks, this would update alternative root to match state of main root. So would be up date if user had updated git via the "refresh git" button in the UI. Either way I have tidied this up now, along with when `alternative_root` is used.

So if ref switching now allowed, `alternative_root` not created as you suggested. And if `missing-dependencies` or `workflow/run` is called without a `ref` then it will use `root` anyway so no need to pull. If they set a `ref`, we check if it is allowed and if so then fetch on `alternative_root ` and use that.